src/mainwin.cc

```diff
@@ -2013,6 +2017,166 @@ void MainWindow::csgRender()
+	bool uploadWorked=uploadStlAndGetPartUrl(exportFilename, userFacingName, partUrl);
+	setCurrentOutput();
+
+	//Check the result:
+	if (! uploadWorked)
+	{
+		PRINT("An error occured while contacting the 3D Print API.");
+		return;
+	}
+	
+	//PRINT("partUrl:");
+	//PRINT(partUrl);
+	
+	//Open the url in the default browser:
+	QDesktopServices::openUrl ( partUrl );
+}
+
+//This function uploads an stl to the 3D printing API endpoint and returns a url that, 
+//when accessed, will show the stl file as a part that can be configured and added to the 
+//shopping cart.  Returns True if successful.
+//
+//Inputs:
+//    exportFilename  - The path to the temporary file that has the stl export in it.
+//    userFacingName  - Then name we should give the file when it is uploaded for the order process.
+//Outputs:
+//    partUrl         - The resulting url to go to next to continue the order process.
+bool MainWindow::uploadStlAndGetPartUrl(const QString & exportFilename, const QString &userFacingName, QUrl &partUrl)
+{
+	setCurrentOutput();
+	
+	//Create a request:
+	QNetworkRequest request(QUrl("https://print.openscad.org/api/v1/part-upload/"));
+	
+	//Set the content header:
+	request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
+	
+	//Get the file name from the file path:
+	QString fileNameBase=QFileInfo(exportFilename).fileName();
+	
+	//Create the request:
+	QJsonObject jsonInput;
+	
+	//Start building the json request:
+	jsonInput.insert("fileName", userFacingName);
+	
+	//Read our stl file:
+	QFile file(exportFilename);
+	if (!file.open(QIODevice::ReadOnly))
+	{
+		PRINT("Unable to open exported stl file.");
+		return false;
+	}
+	
+	//Convert it to base 64:
+	QString stlFileB64=file.readAll().toBase64();
+	
+	//Base 64-encoded file contents:
+	jsonInput.insert("file", stlFileB64);
+	
+	////Pring the input json:
+	//PRINT("Sending this JSON:");
+	//PRINT(QString(QJsonDocument(jsonInput).toJson()).toLocal8Bit().constData());
+	
+	//Create a network access manager:
+	QNetworkAccessManager nam;
+	
+	//Send the post:
+	QNetworkReply *reply = nam.post(request, QJsonDocument(jsonInput).toJson());
+
+	//Wait for the reply:
+	while(!reply->isFinished())
+	{
+		qApp->processEvents();
+	}
+	
+	QVariant statusCodeV = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute);  
+	
+	//Clean up the reply object:
+	reply->deleteLater();
+	
+	char statusStr[250];
+	
+	if (statusCodeV.toInt()!=200)
+	{
+		setCurrentOutput();
+		
+		PRINTB("API status code: %d", statusCodeV.toInt());
+		return false;
+	}
+	
+	//Interpret the response as a json document:
+	QJsonDocument jsonOutDoc= QJsonDocument::fromJson(reply->readAll());
+	
+	//Get the corresponding json object:
+	QJsonObject jsonOutput = jsonOutDoc.object();
+	
+	////Print the whole document:
+	//PRINT(QString(jsonOutDoc.toJson()).toLocal8Bit().constData());
+	
+	//Extract the cartUrl:
+	QString partUrlStr=jsonOutput.value("data").toObject().value("cartUrl").toString();
```

Comment:
What happens if the returned object is malformed, for any reason, e.g. server bug?

Reply:
Then, I assume the user gets an error when we try to go to a url that isn't valid.  What would you like to happen?

Reply:
Hm..... Let me see....

If "jsonOutput.value("data")" is not an object, .toObject returns an empty object:

http://doc.qt.io/qt-5/qjsonvalue.html#toObject-1

> QJsonObject QJsonValue::toObject() const
> This is an overloaded function.
> 
> Converts the value to an object and returns it.
> 
> If type() is not Object, the QJsonObject() will be returned.

Then the .value("cartUrl") will try be applied to the empty QJsonObject:

http://doc.qt.io/qt-5/qjsonobject.html#value

> QJsonValue QJsonObject::value(const QString &key) const
> Returns a QJsonValue representing the value for the key key.
> 
> The returned QJsonValue is QJsonValue::Undefined if the key does not exist.
> 
> See also QJsonValue and QJsonValue::isUndefined().

I tested this out a bit and in that case it seemed to hang, so I added another exception to cover this case.  Thanks!
`    //Start trying to extract the cartUrl:
    auto cartUrlValue=jsonOutput.value("data").toObject().value("cartUrl");

    if (cartUrlValue == QJsonValue::Undefined)
        throw std::runtime_error("Could not get data.cartUrl field from results.");
`